packages/upgrade/src/actions/install.ts

```diff
@@ -0,0 +1,125 @@
+			initial: true,
+		});
+		if (!proceed) {
+			return ctx.exit(0);
+		}
+		
+		log('');
+		
+		await warn('check', `Be sure to follow the ${pluralize('CHANGELOG', majors.length)}.`);
+		for (const pkg of majors.sort(sortPackages)) {
+			await changelog(pkg.name, pkg.changelogTitle!, pkg.changelogURL!);
+		}
+	}
+
+	log('')
+	if (ctx.dryRun) {
+		await info('--dry-run', `Skipping dependency installation`);
+	} else {
+		await runInstallCommand(ctx, dependencies, devDependencies);
+	}
+}
+
+function filterPackages(ctx: Pick<Context, 'packages'>) {
+	const current: PackageInfo[] = [];
+	const dependencies: PackageInfo[] = [];
+	const devDependencies: PackageInfo[] = [];
+	for (const packageInfo of ctx.packages) {
+		const { currentVersion, targetVersion, isDevDependency } = packageInfo;
+		// Remove prefix from `currentVersion` before comparing
+		if (currentVersion.replace(/^\D+/, '') === targetVersion) {
+			current.push(packageInfo);
+		} else {
+			const arr = isDevDependency ? devDependencies : dependencies;
+			arr.push(packageInfo);
+		}
+	}
+	return { current, dependencies, devDependencies }
+}
+
+function sortPackages(a: PackageInfo, b: PackageInfo): number {
+	if (a.isMajor && !b.isMajor) return 1;
+	if (b.isMajor && !a.isMajor) return -1;
+	if (a.name === 'astro') return -1;
+	if (b.name === 'astro') return 1;
+	if (a.name.startsWith('@astrojs') && !b.name.startsWith('@astrojs')) return -1;
+	if (b.name.startsWith('@astrojs') && !a.name.startsWith('@astrojs')) return 1;
+	return a.name.localeCompare(b.name);
+}
+
+async function runInstallCommand(ctx: Pick<Context, 'cwd' | 'packageManager' | 'exit'>, dependencies: PackageInfo[], devDependencies: PackageInfo[]) {
+	const cwd = fileURLToPath(ctx.cwd);
+	if (ctx.packageManager === 'yarn') await ensureYarnLock({ cwd });
+
+	await spinner({
+		start: `Installing dependencies with ${ctx.packageManager}...`,
+		end: `Installed dependencies!`,
+		while: async () => {
+			try {
+				if (dependencies.length > 0) {
+					await shell(ctx.packageManager, ['install', ...dependencies.map(({ name, targetVersion }) => `${name}@${(targetVersion).replace(/^\^/, '')}`)], { cwd, timeout: 90_000, stdio: 'ignore' })
+				}
+				if (devDependencies.length > 0) {
+					await shell(ctx.packageManager, ['install', '--save-dev', ...devDependencies.map(({ name, targetVersion }) => `${name}@${(targetVersion).replace(/^\^/, '')}`)], { cwd, timeout: 90_000, stdio: 'ignore' })
+				}
+			} catch {
+				const packages = [...dependencies, ...devDependencies].map(({ name, targetVersion }) => `${name}@${targetVersion}`).join(' ')
+				log('');
+				error(
+					'error',
+					`Dependencies failed to install, please run the following command manually:\n${color.bold(`${ctx.packageManager} install ${packages}`)}`
+				);
+				return ctx.exit(1);
+			}
+		},
+	});
+
+	await say([`${random(celebrations)} ${random(done)}`, random(bye)], { clear: false });
+}
+
+async function ensureYarnLock({ cwd }: { cwd: string }) {
```

Comment:
What does this function do? The code seems to write something on FS, but the name doesn't hint that. Also, why do we write a lock file? 🤔

Reply:
Added a comment in [f1a16c9](https://github.com/withastro/astro/pull/8525/commits/f1a16c9a99d1d6bb5c411db4dde596c1f590d5bc). This is needed for Yarn PnP support and was added to `create-astro` in https://github.com/withastro/astro/pull/8028